state-chain/traits/src/mocks/cfe_event_emitter_mock.rs

```diff
@@ -0,0 +1,86 @@
+#![cfg(debug_assertions)]
+
+use cf_chains::mocks::{MockEthereum, MockEthereumChainCrypto};
+use codec::{Decode, Encode};
+use frame_support::{storage, StorageHasher, Twox64Concat};
+
+use crate::{CfeEventEmitterForChain, CfeEventEmitterForCrypto, CfeEventEmitterT, Chainflip};
+
+pub struct MockCfeEventEmitter {}
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub enum MockCfeEvent<ValidatorId> {
+	EthThresholdSignatureRequest(
+		cfe_events::ThresholdSignatureRequest<ValidatorId, MockEthereumChainCrypto>,
+	),
+	EthTxBroadcastRequest(cfe_events::TxBroadcastRequest<ValidatorId, MockEthereum>),
+	EthKeygenRequest(cfe_events::KeygenRequest<ValidatorId>),
+	// Note: we don't normally do handover for eth, but this works for tests
+	EthKeyHandoverRequest(cfe_events::KeyHandoverRequest<ValidatorId, MockEthereumChainCrypto>),
+}
+
+const STORAGE_KEY: &[u8] = b"MockCfeEventEmitter::Events";
+
+impl<T: Chainflip> CfeEventEmitterForCrypto<T, MockEthereumChainCrypto> for MockCfeEventEmitter {
+	fn keygen_request(req: cfe_events::KeygenRequest<T::ValidatorId>) {
+		Self::append_event(MockCfeEvent::EthKeygenRequest(req));
+	}
+
+	fn signature_request(
+		req: cfe_events::ThresholdSignatureRequest<T::ValidatorId, MockEthereumChainCrypto>,
+	) {
+		Self::append_event(MockCfeEvent::EthThresholdSignatureRequest(req));
+	}
+
+	fn key_handover_request(
+		req: cfe_events::KeyHandoverRequest<<T as Chainflip>::ValidatorId, MockEthereumChainCrypto>,
+	) {
+		Self::append_event(MockCfeEvent::EthKeyHandoverRequest(req));
+	}
+}
+
+impl<T: Chainflip> CfeEventEmitterForChain<T, MockEthereum> for MockCfeEventEmitter {
+	fn tx_broadcast_request(req: cfe_events::TxBroadcastRequest<T::ValidatorId, MockEthereum>) {
+		Self::append_event(MockCfeEvent::EthTxBroadcastRequest(req));
+	}
+}
+
+impl<T: Chainflip> CfeEventEmitterT<T> for MockCfeEventEmitter {
+	fn peer_registered(
+		_account_id: <T as Chainflip>::ValidatorId,
+		_pubkey: cf_primitives::Ed25519PublicKey,
+		_port: u16,
+		_ip: cf_primitives::Ipv6Addr,
+	) {
+		// TODO: implement when needed for any test
```

Comment:
nit: unimplemented!() might make this easier to catch if/when someone comes to using this

Reply:
This code does get executed though, so the tests would panic. It is not hard to add actual handling, but I'm not sure if it is worth it considering we don't test for these events and we are considering removing peer events in the future.